File: arbolBinarioBusqueda/arbol.cpp
#include "arbol.h"

arbol::arbol() {

	raiz = NULL;
}

bool arbol::estaVacio() {
	return raiz == NULL;
}

void arbol::inserta(int dato) {
	if (estaVacio())
		raiz = new nodo(dato);
	else
		inserta(dato, raiz);
}

void arbol::inserta(int dato, nodo* Nodo) {
	if (dato < Nodo->dato) {
		if (Nodo->izq == NULL)
			Nodo->izq = new nodo(dato);
		else
			inserta(dato, Nodo->izq);
	}
	else {
		if (Nodo->der == NULL)
			Nodo->der = new nodo(dato);
		else
			inserta(dato, Nodo->der);

	}

}

nodo* arbol::buscar(int dato) {
	return buscar(dato, raiz);
}

nodo* arbol::buscar(int dato, nodo* Nodo) {
	if (Nodo == NULL)
		return NULL;

	if (dato == Nodo->dato) 
		return Nodo;
	
	if (dato < Nodo->dato)
		return buscar(dato, Nodo->izq);
	else
		return buscar(dato, Nodo->der);
}

void arbol::mostrarEnPreorden() {
	mostrarEnPreorden(raiz);
}

void arbol::mostrarEnPreorden(nodo* Nodo) {
	if (Nodo == NULL)
		return;
	//mostrar el dato
	cout << Nodo->dato << endl;
	mostrarEnPreorden(Nodo->izq);
	mostrarEnPreorden(Nodo->der);
}

void arbol::mostrarEnOrden() {
	mostrarEnOrden(raiz);
}

void arbol::mostrarEnOrden(nodo* Nodo) {
	if (Nodo == NULL)
		return;
	mostrarEnOrden(Nodo->izq);
	//mostrar el dato
	cout << Nodo->dato << endl;
	mostrarEnOrden(Nodo->der);
}

void arbol::mostrarEnPostorden() {
	mostrarEnPostorden(raiz);
}

void arbol::mostrarEnPostorden(nodo* Nodo) {
	if (Nodo == NULL)
		return;
	mostrarEnPostorden(Nodo->izq);
	mostrarEnPostorden(Nodo->der);
	//mostrar el dato
	cout << Nodo->dato << endl;
}

void arbol::eliminar(int dato) {
	eliminar(dato, raiz, NULL);
}

void arbol::eliminar(int dato, nodo* Nodo, nodo* padre) {
	if (Nodo == NULL) {
		cout << "El nodo especificado no existe" << endl;
		return;
	}

	if (dato < Nodo->dato)
		return eliminar(dato, Nodo->izq, Nodo);
	else if(dato > Nodo->dato)
		return eliminar(dato, Nodo->der, Nodo);


	if (Nodo->izq == NULL && Nodo->der == NULL) {

		if (Nodo == raiz)
			raiz = NULL;
		else {
			if (padre->izq == Nodo)
				padre->izq = NULL;
			else
				padre->der = NULL;
		}
		delete(Nodo);

	}
	else if (Nodo->izq != NULL && Nodo->der == NULL) {
		nodo* aux = Nodo->der;
		nodo* padreaux = Nodo;

		while (aux->izq != NULL) {
			padreaux = aux;
			aux = aux->izq;
		}

		Nodo->dato = aux->dato;
		if (aux->der) {
			padre->izq = aux->der;

			delete(aux);
		}

	}
	else if (Nodo->izq != NULL) {

		if (Nodo == raiz) 
			raiz = Nodo->izq;
		else {
			if (padre->izq == Nodo)
				padre->izq = Nodo->izq;
			else
				padre->der = Nodo->izq;
		}

		delete(Nodo);
	}
	else if (Nodo->der != NULL) {

		if (Nodo == raiz)
			raiz = Nodo->der;
		else {
			if (padre->izq == Nodo)
				padre->izq = Nodo->der;
			else
				padre->der = Nodo->der;
		}

		delete(Nodo);
	}
}

File: Unidad 1/ArbolBinarioJava/Arbol.java
package ArbolBinarioJava;
public class Arbol {
    private Nodo raiz;

    public Arbol() {
        raiz = null;
    }

    public boolean estaVacio() {
        return raiz == null;
    }

    public void insertar(int dato) {
        if (estaVacio())
            raiz = new Nodo(dato);
        else
            insertar(dato, raiz);
    }

    private void insertar(int dato, Nodo Nodo) {
        if (dato < Nodo.getDato()) {
            if (Nodo.getIzq() == null)
                Nodo.setIzq(new Nodo(dato));
            else
                insertar(dato, Nodo.getIzq());
        } else {
            if (Nodo.getDer() == null)
                Nodo.setDer(new Nodo(dato));
            else
                insertar(dato, Nodo.getDer());
        }
    }

    public Nodo buscar(int dato) {
        return buscar(dato, raiz);
    }

    private Nodo buscar(int dato, Nodo Nodo) {
        if (Nodo == null)
            return null;

        if (dato == Nodo.getDato())
            return Nodo;

        if (dato < Nodo.getDato())
            return buscar(dato, Nodo.getIzq());
        else
            return buscar(dato, Nodo.getDer());
    }

    public void mostrarEnPreorden() {
        mostrarEnPreorden(raiz);
    }

    private void mostrarEnPreorden(Nodo Nodo) {
        if (Nodo == null)
            return;

        System.out.println(Nodo.getDato());
        mostrarEnPreorden(Nodo.getIzq());
        mostrarEnPreorden(Nodo.getDer());
    }

    public void mostrarEnOrden() {
        mostrarEnOrden(raiz);
    }

    private void mostrarEnOrden(Nodo Nodo) {
        if (Nodo == null)
            return;

        mostrarEnOrden(Nodo.getIzq());
        System.out.println(Nodo.getDato());
        mostrarEnOrden(Nodo.getDer());
    }

    public void mostrarEnPostorden() {
        mostrarEnPostorden(raiz);
    }

    private void mostrarEnPostorden(Nodo Nodo) {
        if (Nodo == null)
            return;

        mostrarEnPostorden(Nodo.getIzq());
        mostrarEnPostorden(Nodo.getDer());
        System.out.println(Nodo.getDato());
    }

    public void eliminar(int dato) {
        eliminar(dato, raiz, null);
    }

    private void eliminar(int dato, Nodo Nodo, Nodo padre) {
        if (Nodo == null) {
            System.out.println("El nodo especificado no existe");
            return;
        }

        if (dato < Nodo.getDato())
            eliminar(dato, Nodo.getIzq(), Nodo);
        else if (dato > Nodo.getDato())
            eliminar(dato, Nodo.getDer(), Nodo);

        if (Nodo.getIzq() == null && Nodo.getDer() == null) {
            if (Nodo == raiz)
                raiz = null;
            else {
                if (padre.getIzq() == Nodo)
                    padre.setIzq(null);
                else
                    padre.setDer(null);
            }
            Nodo = null;
        } else if (Nodo.getIzq() != null && Nodo.getDer() != null) {
            Nodo aux = Nodo.getDer();
            Nodo padreaux = Nodo;

            while (aux.getIzq() != null) {
                padreaux = aux;
                aux = aux.getIzq();
            }

            Nodo.setDato(aux.getDato());
            if (aux.getDer() != null) {
                padreaux.setIzq(aux.getDer());

                if (padreaux.getIzq() == aux)
                    padreaux.setIzq(null);
                else
                    padreaux.setDer(null);
            }
            aux = null;
        } else if (Nodo.getIzq() != null) {
            if (Nodo == raiz)
                raiz = Nodo.getIzq();
            else {
                if (padre.getIzq() == Nodo)
                    padre.setIzq(Nodo.getIzq());
                else
                    padre.setDer(Nodo.getIzq());
            }
            Nodo = null;
        } else if (Nodo.getDer() != null) {
            if (Nodo == raiz)
                raiz = Nodo.getDer();
            else {
                if (padre.getIzq() == Nodo)
                    padre.setIzq(Nodo.getDer());
                else
                    padre.setDer(Nodo.getDer());
            }
            Nodo = null;
        }
    }
}